libcore/src/main/java/com/mapbox/android/core/crashreporter/MapboxUncaughtExceptionHanlder.java

```diff
@@ -0,0 +1,124 @@
+
+  private MapboxUncaughtExceptionHanlder(Context context,
+                                         String mapboxPackage,
+                                         Thread.UncaughtExceptionHandler defaultExceptionHandler) {
+    this.applicationContext = context;
+    this.mapboxPackage = mapboxPackage;
+    this.defaultExceptionHandler = defaultExceptionHandler;
+
+    SharedPreferences sharedPreferences =
+      context.getSharedPreferences(MAPBOX_CRASH_REPORTER_PREFERENCES, Context.MODE_PRIVATE);
+    sharedPreferences.registerOnSharedPreferenceChangeListener(this);
+  }
+
+  /**
+   * Installs exception handler for particular Mapbox module/sdk
+   * <p>
+   * Crash data will land in context.getFilesDir()/${mapboxPackage}/
+   *
+   * @param context       application context.
+   * @param mapboxPackage mapbox package name exceptions to handle.
+   *                      <p>
+   *                      Note: Package name can be used to control coverage: i.e. `com.mapbox` will catch all
+   *                      mapbox exceptions in the context of a single app process.
+   */
+  public static void install(@Nullable Context context, String mapboxPackage) {
+    if (context == null) {
+      Log.i(TAG, "Uncaught exception handler cannot be installed: context == null");
+      return;
+    }
+
+    Thread.setDefaultUncaughtExceptionHandler(new MapboxUncaughtExceptionHanlder(context,
+      mapboxPackage, Thread.getDefaultUncaughtExceptionHandler()));
+  }
+
+  @Override
+  public void uncaughtException(Thread t, Throwable e) {
+    // If we're not enabled or crash is not in Mapbox code
+    // then just pass the Exception on to the defaultExceptionHandler.
+    if (isEnabled.get() && isMapboxCrash(e)) {
+      // Handle Mapbox exception
+      try {
+        // TODO: capture crash data
+        String timestamp = "";
+        String json = "";
+
+        // TODO: create file
+        File file = FileUtils.getFile(applicationContext,
+          getReportFileName(mapboxPackage, timestamp));
+
+        if (!file.exists()) {
+          // Create file
+        }
+
+        // TODO: write data to disk
+        FileUtils.writeToFile(file, json);
+      } catch (Exception ex) {
+        Log.e(TAG, ex.toString());
+      }
+    }
+
+    // Give default exception handler a chance to handle exception
+    if (defaultExceptionHandler != null) {
+      defaultExceptionHandler.uncaughtException(t, e);
+    } else {
+      Log.i(TAG, "Default exception handler is null");
+    }
+  }
+
+  private boolean isMapboxCrash(@Nullable Throwable th) {
+    Throwable cause = th;
+    while (cause != null) {
+      final StackTraceElement[] stackTraceElements = cause.getStackTrace();
+      for (final StackTraceElement e : stackTraceElements) {
+        String className = e.getClassName();
+        if (className.startsWith(mapboxPackage)) {
+          return true;
+        }
+      }
+      cause = cause.getCause();
+    }
+    return false;
```

Comment:
This will capture any crash that has some kind of mapbox package in the stacktrace. This however can also result in capturing not Mapbox related crashes. For example, a user makes a null pointer exception inside the OnMapReady callback with code unrelated to our SDKs. Afaik with above implementation this is flagged as a mapbox crash.

Reply:
@tobrun fair concern! this is a naive implementation, which i plan to generalize/improve as i dive deeper into extensive integration testing. based on my preliminary testing and given that telemetry sdk literally exposes 3 public apis (which i hope we can reduce to 1 in future), it sort of works :)